---
title: "Pertussis Mini Project"
author: "Ethan Lai"
date: "3/9/2022"
output: pdf_document
---
Q1. With the help of the R “addin” package datapasta assign the CDC pertussis case number data to a data frame called cdc and use ggplot to make a plot of cases numbers over time.

```{r}
cdc<-data.frame(
                               Year = c(1922L,1923L,1924L,1925L,1926L,
                                        1927L,1928L,1929L,1930L,1931L,1932L,
                                        1933L,1934L,1935L,1936L,1937L,1938L,
                                        1939L,1940L,1941L,1942L,1943L,
                                        1944L,1945L,1946L,1947L,1948L,1949L,
                                        1950L,1951L,1952L,1953L,1954L,1955L,
                                        1956L,1957L,1958L,1959L,1960L,
                                        1961L,1962L,1963L,1964L,1965L,1966L,
                                        1967L,1968L,1969L,1970L,1971L,1972L,
                                        1973L,1974L,1975L,1976L,1977L,1978L,
                                        1979L,1980L,1981L,1982L,1983L,
                                        1984L,1985L,1986L,1987L,1988L,1989L,
                                        1990L,1991L,1992L,1993L,1994L,1995L,
                                        1996L,1997L,1998L,1999L,2000L,
                                        2001L,2002L,2003L,2004L,2005L,2006L,
                                        2007L,2008L,2009L,2010L,2011L,2012L,
                                        2013L,2014L,2015L,2016L,2017L,2018L,
                                        2019L),
       No..Reported.Pertussis.Cases = c(107473,164191,165418,152003,
                                        202210,181411,161799,197371,166914,
                                        172559,215343,179135,265269,180518,
                                        147237,214652,227319,103188,183866,
                                        222202,191383,191890,109873,133792,
                                        109860,156517,74715,69479,120718,68687,
                                        45030,37129,60886,62786,31732,28295,
                                        32148,40005,14809,11468,17749,
                                        17135,13005,6799,7717,9718,4810,3285,
                                        4249,3036,3287,1759,2402,1738,
                                        1010,2177,2063,1623,1730,1248,1895,
                                        2463,2276,3589,4195,2823,3450,4157,
                                        4570,2719,4083,6586,4617,5137,
                                        7796,6564,7405,7298,7867,7580,9771,
                                        11647,25827,25616,15632,10454,13278,
                                        16858,27550,18719,48277,28639,
                                        32971,20762,17972,18975,15609,18617)
     )

```

```{r}
colnames(cdc)
```


```{r}

library(ggplot2)
cdcplot<- ggplot(cdc,aes(x=Year, y=No..Reported.Pertussis.Cases)) + geom_point() + geom_line()
cdcplot

```
Q2. Using the ggplot geom_vline() function add lines to your previous plot for the 1946 introduction of the wP vaccine and the 1996 switch to aP vaccine (see example in the hint below). What do you notice?


```{r}
vaxYears <- c(1946, 1996)
cdcplot + geom_vline(xintercept = vaxYears, linetype="dashed", col=c("red", "blue"))
```

Q3. Describe what happened after the introduction of the aP vaccine? Do you have a possible explanation for the observed trend?

There seems to be a mild resurgence in pertussis rates. Maybe the aP vaccine is somewhat less effective? Or anti-vaxxers emerged around then...

```{r}
library("jsonlite")
subject <- read_json("https://www.cmi-pb.org/api/subject", simplifyVector = TRUE) 
head(subject)
```
Q4. How many aP and wP infancy vaccinated subjects are in the dataset?

```{r}
table(subject$infancy_vac)

```
49 wP subjects, 47 aP subjects.

Q5. How many Male and Female subjects/patients are in the dataset?

```{r}
table(subject$biological_sex)
```
66 female subjects, 30 male subjects

Q6. What is the breakdown of race and biological sex (e.g. number of Asian females, White males etc…)?:
```{r}
table(subject$race)
```


```{r}
specimen <- read_json("https://www.cmi-pb.org/api/specimen", simplifyVector = TRUE) 
titer <- read_json("https://www.cmi-pb.org/api/ab_titer", simplifyVector = TRUE) 

library(dplyr)
```

Now to make sense of this data and ask questions about ap vs wP of the Ab titer data. We need to join subject with these new tables.
```{r}
meta <- inner_join(subject, specimen)
dim(meta)
```

```{r}
abdata<- inner_join (titer, meta)
dim (abdata)
```

```{r}
head(abdata)
```
Q11. How many specimens (i.e. entries in abdata) do we have for each isotype?

```{r}
table(abdata$isotype)
```

Q12. What do you notice about the number of visit 8 specimens compared to other visits?

```{r}
table(abdata$visit)
```
There are far, far fewer visit 8s compared to the others.
So, let's filter it out for later analysis

```{r}
ig1 <- abdata %>% filter(isotype == "IgG1", visit!=8)
head(ig1)
```


Q13. Complete the following code to make a summary boxplot of Ab titer levels for all antigens:

```{r}
ggplot(ig1) +
  aes(ab_titer, antigen) +
  geom_boxplot() + 
  facet_wrap(vars(visit), nrow=3)
```


```{r}
filter(ig1, antigen=="Measles") %>%
  ggplot() +
  aes(ab_titer, col=infancy_vac) +
  geom_boxplot(show.legend = TRUE) +
  facet_wrap(vars(visit)) +
  theme_bw()
```

```{r}
filter(ig1, antigen=="FIM2/3") %>%
  ggplot() +
  aes(ab_titer, col=infancy_vac) +
  geom_boxplot(show.legend = TRUE) +
  facet_wrap(vars(visit)) +
  theme_bw()
```
Q16. What do you notice about these two antigens time courses and the FIM2/3 data in particular?

Measles doesn't change much over time, while Fib2/3, as expected, increases dramatically over the visit time course.

Q17. Do you see any clear difference in aP vs. wP responses?

Nothing very significant.


For RNA-Seq data the API query mechanism quickly hits the web browser interface limit for file size. We will present alternative download mechanisms for larger CMI-PB datasets in the next section. However, we can still do “targeted” RNA-Seq querys via the web accessible API.

For example we can obtain RNA-Seq results for a specific ENSEMBLE gene identifier or multiple identifiers combined with the & character:



```{r}
url <- "https://www.cmi-pb.org/api/v2/rnaseq?versioned_ensembl_gene_id=eq.ENSG00000211896.7"

rna <- read_json(url, simplifyVector = TRUE)
```

```{r}
#meta <- inner_join(specimen, subject)
ssrna <- inner_join(rna, meta)
head(ssrna)
```
Q18. Make a plot of the time course of gene expression for IGHG1 gene (i.e. a plot of visit vs. tpm).

```{r}
ggplot(ssrna) +
  aes(visit, raw_count, group=subject_id, color = infancy_vac) +
  geom_point() +
  geom_line(alpha=0.2)
```
Q19.: What do you notice about the expression of this gene (i.e. when is it at it’s maximum level)?

UsuallyvVisit 4.

Q20. Does this pattern in time match the trend of antibody titer data? If not, why not?

Not entirely. Antibody titer data peaks somewhat later, around visit 5, by which point gene expression level has already dropped. THis makes sense, as we'd expect a lag between transcription and translation, and we'd expect antibodies to persist, leading to a later peak.
